internal/provider/waypoint/data_source_waypoint_action.go

```diff
@@ -0,0 +1,207 @@
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Data Source Configure Type",
+			fmt.Sprintf("Expected *clients.Client, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+		return
+	}
+	d.client = client
+}
+
+func (d *DataSourceAction) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var data ActionResourceModel
+
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+
+	client := d.client
+	if d.client == nil {
+		resp.Diagnostics.AddError(
+			"Unconfigured HCP Client",
+			"Expected configured HCP client. Please report this issue to the provider developers.",
+		)
+		return
+	}
+
+	projectID := client.Config.ProjectID
+	if !data.ProjectID.IsNull() {
+		projectID = data.ProjectID.ValueString()
+	}
+
+	loc := &sharedmodels.HashicorpCloudLocationLocation{
+		OrganizationID: client.Config.OrganizationID,
+		ProjectID:      projectID,
+	}
+
+	var actionModel *waypoint_models.HashicorpCloudWaypointActionConfig
+	var err error
+
+	actionModel, err = clients.GetAction(ctx, client, loc, data.ID.ValueString(), data.Name.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddError(err.Error(), "Failed to find action by ID or name")
+		return
+	}
+
+	if actionModel.ID != "" {
+		data.ID = types.StringValue(actionModel.ID)
+	}
+	if actionModel.Name != "" {
+		data.Name = types.StringValue(actionModel.Name)
+	}
+	if actionModel.Description != "" {
+		data.Description = types.StringValue(actionModel.Description)
+	}
+
+	data.OrgID = types.StringValue(client.Config.OrganizationID)
+	data.ProjectID = types.StringValue(client.Config.ProjectID)
+
+	data.Request = &actionRequest{}
+	headerMap := make(map[string]string)
+
+	var diags diag.Diagnostics
+
+	// In the future, expand this to accommodate other types of requests
+
+	data.Request.Custom = &customRequest{}
+	if actionModel.Request.Custom.Method != nil {
+		methodString, err := convertMethodToStringType(*actionModel.Request.Custom.Method)
+		if err != nil {
+			resp.Diagnostics.AddError(
+				"Unexpected HTTP Method",
+				"Expected GET, POST, PUT, DELETE, or PATCH. Please report this issue to the provider developers.",
+			)
+			return
+		} else {
+			data.Request.Custom.Method = methodString
+		}
+	}
+	if actionModel.Request.Custom.Headers != nil {
+		for _, header := range actionModel.Request.Custom.Headers {
+			headerMap[header.Key] = header.Value
+		}
+		if len(headerMap) > 0 {
+			data.Request.Custom.Headers, diags = types.MapValueFrom(ctx, types.StringType, headerMap)
+			resp.Diagnostics.Append(diags...)
+			if resp.Diagnostics.HasError() {
+				return
+			}
+		} else {
+			data.Request.Custom.Headers = types.MapNull(types.StringType)
+		}
+	}
+	if actionModel.Request.Custom.URL != "" {
+		data.Request.Custom.URL = types.StringValue(actionModel.Request.Custom.URL)
+	}
+	if actionModel.Request.Custom.Body != "" {
+		data.Request.Custom.Body = types.StringValue(actionModel.Request.Custom.Body)
```

Comment:
same comment here about always setting the value in state to whatever we get from the API